qualtran/bloqs/basic_gates/z_basis.py

```diff
@@ -230,15 +230,26 @@ def signature(self) -> Signature:
         side = Side.RIGHT if self.state else Side.LEFT
         return Signature([Register('val', bitsize=self.bitsize, side=side)])
 
-    def build_composite_bloq(self, bb: 'BloqBuilder') -> Dict[str, 'SoquetT']:
-        states = [ZeroState(), OneState()]
-        xs = []
-        for bit in ints_to_bits(np.array([self.val]), w=self.bitsize)[0]:
-            x = bb.add(states[bit])
-            xs.append(x)
-        xs = np.array(xs)
-
-        return {'val': bb.join(xs)}
+    def build_composite_bloq(self, bb: 'BloqBuilder', **val) -> Dict[str, 'SoquetT']:
+        bits = ints_to_bits(np.array([self.val]), w=self.bitsize)[0]
+
+        if self.state:
+            assert not val
+            states = [ZeroState(), OneState()]
+            xs = []
+            for bit in bits:
+                x = bb.add(states[bit])
+                xs.append(x)
+            xs = np.array(xs)
+
+            return {'val': bb.join(xs)}
+
+        val = val['val']
+        xs = bb.split(val)
+        effects = [ZeroEffect(), OneEffect()]
+        for i, bit in enumerate(bits):
+            bb.add(effects[bit], q=xs[i])
+        return {}
```

Comment:
I have a small preference to split the two decompositions into `_build_composite_bloq_effect` and `_build_composite_bloq_state` and then call either of the two methods using an if / else on `self.state` from `_build_composite_bloq`. 

I know maybe initially I proposed that we should have a common base class for `_IntVector` but at this point, I'm also fine if we want to split it into `_IntEffect` and `_IntState`.

Reply:
I did the refactor